elements/lisk-p2p/src/peer_directory/new_list.ts

```diff
@@ -50,54 +50,54 @@ export class NewList extends PeerList {
 	}
 
 	// Extend eviction of NewPeers
-	public evictPeer(bucketId: number): CustomPeerInfo {
+	public evictPeers(bucketId: number): ReadonlyArray<CustomPeerInfo> {
```

Comment:
Actually yes, removing only one is better. If all the peers just happen to all expire at the same time it would be a problem.

Reply:
Yes, it will be weird if one new peer deletes all the peers in a bucket if all of them are 30 days older. It should be a peer for a peer strategy.

Reply:
`Any address` means to me to expire every address. Once you start cleaning a bucket, I'd clean it completely. Please @janhack, could you shed us some light here?

Reply:
The main purpose of the `newPeers` collection is to store a reasonable number of possible peers (<=32 per bucket, <=4096 in total). This collection should not grow over these limits (to avoid spamming of peers or poising the `newPeers` with addresses controlled by an adversary), but also not to small so a node has sufficient peers to try as backup. I think the eviction should only evict peers if necessary (i.e., the bucket is full) and only as many peers as necessary (only one peer if one peer needs to be added).

A different topic is, if we should maybe introduce some `newPeers` housekeeping. This could mean that we go over all addresses in `newPeers`, make a connection attempt (to ensure that there is actually a Lisk Node at this address) and update a timestamp if the attempt is successful. As part of this housekeeping we could also remove peers if they have not been reachable for a certain period of time. I would think that it would be better to do this separately from the peer eviction.